Source/DafnyCore/Substituter.cs

```diff
@@ -930,7 +979,73 @@ public class Substituter {
+    /// Substitution in a comprehension expression. 
+    /// Parameter 'forceSubstituteOfBoundVars' should be set to false if and only if
+    /// the expression is a binding guard, in which case a bound variable is introduced.
+    /// Such a variable must not be substituted. 
+    /// </summary>
+    protected Expression SubstituteComprehensionExpr(ComprehensionExpr expr, bool forceSubstituteOfBoundVars) {
+
+      Expression newExpr = null;
+
+      var e = (ComprehensionExpr)expr;
+      // For quantifiers and setComprehesion we want to make sure that we don't introduce name clashes with
+      // the enclosing scopes.
+
+      var q = e as QuantifierExpr;
+      if (q != null && q.SplitQuantifier != null) {
+        if (forceSubstituteOfBoundVars) {
+          return Substitute(q.SplitQuantifierExpression);
+        } else {
+          return SubstituteComprehensionExpr((ComprehensionExpr)q.SplitQuantifierExpression, false);
+        }
+      }
+
+      var newBoundVars = CreateBoundVarSubstitutions(e.BoundVars,
+        forceSubstituteOfBoundVars && (expr is ForallExpr || expr is ExistsExpr || expr is SetComprehension));
+      var newRange = e.Range == null ? null : Substitute(e.Range);
+      var newTerm = Substitute(e.Term);
+      var newAttrs = SubstAttributes(e.Attributes);
+      var newBounds = SubstituteBoundedPoolList(e.Bounds);
+      if (newBoundVars != e.BoundVars || newRange != e.Range || newTerm != e.Term || newAttrs != e.Attributes ||
+          newBounds != e.Bounds || !forceSubstituteOfBoundVars) {
+        if (e is SetComprehension) {
+          newExpr = new SetComprehension(e.BodyStartTok, e.BodyEndTok, ((SetComprehension)e).Finite, newBoundVars,
+            newRange, newTerm, newAttrs);
+        } else if (e is MapComprehension) {
+          var mc = (MapComprehension)e;
+          var newTermLeft = mc.IsGeneralMapComprehension ? Substitute(mc.TermLeft) : null;
+          newExpr = new MapComprehension(e.BodyStartTok, e.BodyEndTok, mc.Finite, newBoundVars, newRange, newTermLeft,
+            newTerm, newAttrs);
+        } else if (expr is ForallExpr forallExpr) {
+          newExpr = new ForallExpr(expr.tok, e.BodyEndTok, newBoundVars, newRange, newTerm, newAttrs);
+        } else if (expr is ExistsExpr existsExpr) {
+          newExpr = new ExistsExpr(expr.tok, e.BodyEndTok, newBoundVars, newRange, newTerm, newAttrs);
+        } else if (expr is LambdaExpr) {
+          var l = (LambdaExpr)expr;
+          newExpr = new LambdaExpr(e.BodyStartTok, e.BodyEndTok, newBoundVars, newRange,
+            l.Reads.ConvertAll(SubstFrameExpr), newTerm);
+        } else {
+          Contract.Assert(false); // unexpected ComprehensionExpr
+        }
+
+        ((ComprehensionExpr)newExpr).Bounds = newBounds;
+      }
+
+      // undo any changes to substMap (could be optimized to do this only if newBoundVars != e.BoundVars)
+      foreach (var bv in e.BoundVars) {
+        substMap.Remove(bv);
+      }
+
+      return newExpr;
```

Comment:
Could we add an assertion before this to say when `newExpr` can be null?
I think it's something like this?

```
Contract.Assert(!forceSubstituteOfBoundVars ==> newExpr != null)
```

I was confused when I first read through the code to see that this was used directly in the `if` case but that it may also return null.

Reply:
We could but here is my concern: it is not actually a required invariant, it just happens to be implemented that way to match the rest of the code.